File: app/src/main/java/com/timsimonhughes/atlas/views/PageIndicatorView.java
package com.timsimonhughes.atlas.views;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

import com.timsimonhughes.atlas.R;

import androidx.annotation.ColorRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.viewpager.widget.ViewPager;

public class PageIndicatorView extends View {

    private static final int DEFAULT_CHOSEN_CIRCLE_COLOR = Color.parseColor("#ffffff");
    private static final int DEFAULT_CIRCLE_COLOR = Color.parseColor("#000000");
    private static final int DEFAULT_DISTANCE_BETWEEN_CIRCLES_IN_DP = 5, DEFAULT_CIRCLE_RADIUS_IN_DP = 5, DEFAULT_NUMBER_OF_CIRCLES = 0;

    private Paint inactiveCirclePaint, activeCirclePaint;

    private float circleRadiusInPX;
    private float distanceBetweenCircleInPX;

    private float[] circleLocations;
    private float yLocation;

    private int activeCircleColor;
    private int inactiveCircleColor;
    private int numberOfCircles;

    private int currentIndex = 0;

//    private LifeCycleObserver lifeCycleObserver;

    public PageIndicatorView(Context context) {
        super(context);
        init(context, null);
    }

    public PageIndicatorView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    public PageIndicatorView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs);
    }

    @TargetApi(21)
    public PageIndicatorView(Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(context, attrs);
    }

    private void init(Context context, AttributeSet attributes) {

        if (attributes != null) {
            TypedArray attributeArray = context.obtainStyledAttributes(attributes, R.styleable.PageIndicatorView);

            circleRadiusInPX =
                    attributeArray.getDimension(
                            R.styleable.PageIndicatorView_circle_radius,
                            dpToPx(context, DEFAULT_CIRCLE_RADIUS_IN_DP));

            distanceBetweenCircleInPX =
                    attributeArray.getDimension(
                            R.styleable.PageIndicatorView_distance_between_circles,
                            dpToPx(context, DEFAULT_DISTANCE_BETWEEN_CIRCLES_IN_DP));

            numberOfCircles =
                    attributeArray.getInt(
                            R.styleable.PageIndicatorView_number_of_circles,
                            DEFAULT_NUMBER_OF_CIRCLES);

            activeCircleColor =
                    attributeArray.getColor(
                            R.styleable.PageIndicatorView_active_circle_color,
                            DEFAULT_CHOSEN_CIRCLE_COLOR);

            inactiveCircleColor =
                    attributeArray.getColor(
                            R.styleable.PageIndicatorView_inactive_circle_color,
                            DEFAULT_CIRCLE_COLOR);

            attributeArray.recycle();

//            activeCircleRadius = circleRadiusInPX * 2;

        } else {

            circleRadiusInPX = dpToPx(context, DEFAULT_CIRCLE_RADIUS_IN_DP);
            distanceBetweenCircleInPX = dpToPx(context, DEFAULT_DISTANCE_BETWEEN_CIRCLES_IN_DP);
            numberOfCircles = DEFAULT_NUMBER_OF_CIRCLES;
            activeCircleColor = DEFAULT_CHOSEN_CIRCLE_COLOR;
            inactiveCircleColor = DEFAULT_CIRCLE_COLOR;
        }

        if (inactiveCirclePaint == null) {
            inactiveCirclePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
            inactiveCirclePaint.setColor(inactiveCircleColor);
            inactiveCirclePaint.setStyle(Paint.Style.FILL);
        }

        if (activeCirclePaint == null) {
            activeCirclePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
            activeCirclePaint.setColor(activeCircleColor);
            activeCirclePaint.setStyle(Paint.Style.FILL);
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        if (circleLocations != null && circleLocations.length != 0) {
            for (int i = 0; i < numberOfCircles; ++i) {
                if (i == currentIndex) {
                    canvas.drawCircle(circleLocations[i], yLocation, circleRadiusInPX, activeCirclePaint);
                } else {
                    canvas.drawCircle(circleLocations[i], yLocation, circleRadiusInPX, inactiveCirclePaint);
                }
            }
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        boolean isMatchParentFlagOn = MeasureSpec.getMode(widthMeasureSpec) == MeasureSpec.EXACTLY;
        measureIndicesOfIndicatorStations(isMatchParentFlagOn);

        int measuredWidth = numberOfCircles > 0 ? measureWidth(widthMeasureSpec) : 0;
        int measuredHeight = numberOfCircles > 0 ? measureHeight(heightMeasureSpec) : 0;

        yLocation = measuredHeight / 2;

        setMeasuredDimension(measuredWidth, measuredHeight);
    }

    private int measureHeight(int measureSpec) {

        int specMode = MeasureSpec.getMode(measureSpec);
        int specSize = MeasureSpec.getSize(measureSpec);
        int result;
        if (specMode == MeasureSpec.EXACTLY) {
            result = specSize;
        } else {
            result = (int) (circleRadiusInPX * 2 + getPaddingTop() + getPaddingBottom());
        }

        return result;
    }

    private int measureWidth(int measureSpec) {

        int specMode = MeasureSpec.getMode(measureSpec);
        int specSize = MeasureSpec.getSize(measureSpec);
        int result;
        if (specMode == MeasureSpec.EXACTLY) {
            result = specSize;
        } else {
            result = (int) (circleLocations[numberOfCircles-1] + circleRadiusInPX + getPaddingRight());
        }

        return result;
    }

    private void measureIndicesOfIndicatorStations(boolean shouldAddScreenOffset) {
        if (numberOfCircles > 0 && circleLocations == null) {
            circleLocations = new float[numberOfCircles];

            float xCounter = getPaddingLeft() + circleRadiusInPX;
            float spaceBetweenCircles = circleRadiusInPX * 2 + distanceBetweenCircleInPX;

            for (int index = 0; index < numberOfCircles; index++) {
                circleLocations[index] = xCounter;

                xCounter += spaceBetweenCircles;
            }

            if (shouldAddScreenOffset) {
                int screenWidth = getContext().getResources().getDisplayMetrics().widthPixels;
                int widthOfView = (int) (circleLocations[numberOfCircles-1] + circleRadiusInPX + getPaddingRight());
                int spaceToAddFromEachSide = (screenWidth - widthOfView) / 2;

                for (int index = 0; index < circleLocations.length; index++) {
                    circleLocations[index] += spaceToAddFromEachSide;
                }
            }
        }
    }

    @Override
    public void requestLayout() {
        circleLocations = null;
        super.requestLayout();
    }

//    @Override
//    protected void onDetachedFromWindow() {
//        if (lifeCycleObserver != null) {
//            lifeCycleObserver.onDetached();
//            lifeCycleObserver = null;
//        }
//
//        super.onDetachedFromWindow();
//    }

    private static float dpToPx(final Context context, final float value) {
        return value * context.getResources().getDisplayMetrics().density;
    }

    public void setCurrentIndex(int index) {
        this.currentIndex = index;
        invalidate();
    }

    int getNumberOfCircles() {
        return numberOfCircles;
    }

    public void setNumberOfCircles(int numberOfCircles) {
        this.numberOfCircles = numberOfCircles;
    }

    public void setChosenCircleColor(@NonNull Context context, @ColorRes int color) {
        this.activeCirclePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        this.activeCirclePaint.setColor(ContextCompat.getColor(context, color));
        this.activeCirclePaint.setStyle(Paint.Style.FILL);
    }

    public void setActiveCircleColor() {
        // TODO: Logic to add active circle color
    }

    public void setInactiveCircleColor() {
        // TODO: Logic to add inactive circle color
    }

    public void setDefaultCircleColor(@NonNull Context context, @ColorRes int color) {
        this.inactiveCirclePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        this.inactiveCirclePaint.setColor(ContextCompat.getColor(context, color));
        this.inactiveCirclePaint.setStyle(Paint.Style.FILL);
    }

    public void setCircleRadiusInDp(@NonNull Context context, float circleRadiusInDp) {
        this.circleRadiusInPX = dpToPx(context, circleRadiusInDp);
    }

    public void setDistanceBetweenCircleInDp(@NonNull Context context, float distanceBetweenCircleInPX) {
        this.distanceBetweenCircleInPX = dpToPx(context, distanceBetweenCircleInPX);
    }

    public void setupWithViewpager(ViewPager viewPager) {
        this.currentIndex = viewPager.getCurrentItem();
    }
}

File: appcore/src/main/java/com/timsimonhughes/appcore/views/pageIndicator/PageIndicatorView.kt
package com.timsimonhughes.appcore.views.pageIndicator

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.os.Parcel
import android.os.Parcelable
import android.util.AttributeSet
import android.view.View
import androidx.viewpager2.widget.ViewPager2
import com.timsimonhughes.appcore.R
import com.timsimonhughes.appcore.views.cardStackView.CardStackLayoutManager

class PageIndicatorView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet?,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    private val defaultActiveIndicatorColor = Color.parseColor("#ffffff")
    private val defaultInactiveIndicatorColor = Color.parseColor("#000000")
    private val defaultDistanceBetweenIndicators = 5
    private val defaultNumberOfIndicators = 3
    private val defaultIndicatorRadiusDp = 5

    private var activeIndicatorPaint: Paint? = null
    private var inactiveIndicatorPaint: Paint? = null

    enum class IndicatorShape { Circle, RoundedRect }

    private var indicatorRadiusInPx: Float = 0f
    private var distanceBetweenIndicatorsInPx: Float = 0f

    private var circleLocations: FloatArray? = null
    private var yLocation: Float? = null

    private var activeIndicatorColor: Int? = null
    private var inactiveIndicatorColor: Int? = null
    private var numberOfIndicators: Int = 2

    private var currentIndex: Int = 0

    init {
        attrs?.let {
            val typedArray = context.obtainStyledAttributes(attrs, R.styleable.PageIndicatorView)
            indicatorRadiusInPx = typedArray.getDimension(
                R.styleable.PageIndicatorView_circle_radius,
                dpToPx(context, defaultIndicatorRadiusDp.toFloat())
            )
            distanceBetweenIndicatorsInPx = typedArray.getDimension(
                R.styleable.PageIndicatorView_distance_between_circles,
                defaultDistanceBetweenIndicators.toFloat()
            )
            numberOfIndicators = typedArray.getInt(
                R.styleable.PageIndicatorView_number_of_circles,
                defaultNumberOfIndicators
            )
            activeIndicatorColor = typedArray.getColor(
                R.styleable.PageIndicatorView_active_circle_color,
                defaultActiveIndicatorColor
            )
            inactiveIndicatorColor = typedArray.getColor(
                R.styleable.PageIndicatorView_inactive_circle_color,
                defaultInactiveIndicatorColor
            )
            typedArray.recycle()
        }

        if (inactiveIndicatorPaint == null) {
            inactiveIndicatorPaint = Paint().apply {
                isAntiAlias = true
                color = inactiveIndicatorColor!!
                style = Paint.Style.FILL
            }
        }

        if (activeIndicatorPaint == null) {
            activeIndicatorPaint = Paint().apply {
                isAntiAlias = true
                color = activeIndicatorColor!!
                style = Paint.Style.FILL
            }
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val isMatchParent = MeasureSpec.getMode(widthMeasureSpec) == MeasureSpec.EXACTLY
        measureIndicators(isMatchParent)

        val measuredWidth = if (numberOfIndicators > 0) measureWidth(widthMeasureSpec) else 0
        val measuredHeight = if (numberOfIndicators > 0) measureHeight(heightMeasureSpec) else 0

        yLocation = (measuredHeight / 2).toFloat()

        setMeasuredDimension(measuredWidth, measuredHeight)
    }

    private fun measureHeight(measureSpec: Int): Int {
        val specMode = MeasureSpec.getMode(measureSpec)
        val specSize = MeasureSpec.getSize(measureSpec)

        return if (specMode == MeasureSpec.EXACTLY) {
            specSize
        } else {
            (indicatorRadiusInPx * 2 + paddingTop + paddingBottom).toInt()
        }
    }

    private fun measureWidth(measureSpec: Int) : Int {
        val specMode = MeasureSpec.getMode(measureSpec)
        val specSize = MeasureSpec.getSize(measureSpec)

        return if (specMode == MeasureSpec.EXACTLY) {
            specSize
        } else {
            (circleLocations!![numberOfIndicators - 1] + indicatorRadiusInPx + paddingRight).toInt()
        }
    }

    private fun measureIndicators(isMatchParent: Boolean) {
        if (numberOfIndicators > 0 && circleLocations == null) {

            circleLocations = FloatArray(numberOfIndicators)

            var xCounter = paddingLeft + indicatorRadiusInPx
            val spaceBetweenIndicators = indicatorRadiusInPx * 2 + distanceBetweenIndicatorsInPx

            for (i in 0 until numberOfIndicators) {
                circleLocations!![i] = xCounter
                xCounter += spaceBetweenIndicators
            }

            if (isMatchParent) {
                val screenWidth = context.resources.displayMetrics.widthPixels
                val widthOfView =
                    (circleLocations!![numberOfIndicators - 1] + indicatorRadiusInPx + paddingRight)
                val spaceToAddFromEachSide = (screenWidth - widthOfView) / 2

                for (i in circleLocations!!.indices) {
                    circleLocations!![i] += spaceToAddFromEachSide
                }
            }
        }
    }

    override fun requestLayout() {
        circleLocations = null
        super.requestLayout()
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        if (circleLocations != null) {
            for (i in 0 until numberOfIndicators) {
                if (i == currentIndex) {
                    canvas?.drawCircle(
                        circleLocations!![i],
                        yLocation!!,
                        indicatorRadiusInPx,
                        activeIndicatorPaint!!
                    )
                } else {
                    canvas?.drawCircle(
                        circleLocations!![i],
                        yLocation!!,
                        indicatorRadiusInPx,
                        inactiveIndicatorPaint!!
                    )
                }
            }
        }
    }

    override fun onSaveInstanceState(): Parcelable? {
        val superState = super.onSaveInstanceState()
        val savedState = SavedState(superState)
        savedState.index = this.currentIndex

        // Return our state along with our super class's state.
        return savedState
    }

    override fun onRestoreInstanceState(state: Parcelable?) {
        // Cast the incoming Parcelable to our custom SavedState. We produced
        // this Parcelable before, so we know what type it is.

        val savedState = state as SavedState

        // Let our super class process state before we do because we should
        // depend on our super class, we shouldn't imply that our super class
        // might need to depend on us.

        super.onRestoreInstanceState(savedState.superState)

        // Grab our properties out of our SavedState.

        this.currentIndex = savedState.index
        // Update our visuals in whatever way we want, like...

        requestLayout()

        invalidate()
    }

    class SavedState : BaseSavedState {
        var index = 0

        constructor(superState: Parcelable?) : super(superState)

        constructor(source: Parcel) : super(source) {
            index = source.readInt()
        }

        override fun writeToParcel(out: Parcel, flags: Int) {
            super.writeToParcel(out, flags)
            out.writeInt(index)
        }

        val CREATOR = object : Parcelable.Creator<SavedState> {
            override fun createFromParcel(source: Parcel): SavedState {
                return SavedState(source)
            }

            override fun newArray(size: Int): Array<SavedState?> {
                return arrayOfNulls(size)
            }
        }
    }

    /**
     * Helper methods
     **/

    private fun dpToPx(context: Context, value: Float): Float {
        return value * context.resources.displayMetrics.density
    }

    /**
     * Get/Set methods
     **/

    fun getCurrentIndex(index: Int) {
        this.currentIndex = index
        invalidate()
    }

    fun getNumberOfIndicators(): Int {
        return numberOfIndicators
    }

    fun setNumberOfIndicators(indicatorCount: Int) {
        this.numberOfIndicators = indicatorCount
    }

    fun setCircleRadius(indicatorRadius: Float) {
        this.indicatorRadiusInPx = dpToPx(context, indicatorRadius)
    }

    fun setDistanceBetweenIndicators(distanceBetweenIndicators: Float) {
        this.distanceBetweenIndicatorsInPx = dpToPx(context, distanceBetweenIndicators)
    }

    fun setActiveIndicatorColor(activeColor: Int) {
        this.activeIndicatorColor = activeColor
    }

    fun setInactiveIndicatorColor(inactiveColor: Int) {
        this.inactiveIndicatorColor = inactiveColor
    }

    fun setupWithViewPager(viewpager: ViewPager2) {
        this.currentIndex = viewpager.currentItem
        invalidate()
    }

    fun setupWithRecyclerView(layoutManager: CardStackLayoutManager) {
        this.numberOfIndicators = layoutManager.itemCount
        this.currentIndex = layoutManager.getTopPosition()
        invalidate()
    }

    fun setCurrentIndex(index: Int) {
        this.currentIndex = index
        invalidate()
    }
}